cpp/src/arrow/io/memory.h

```diff
@@ -24,12 +24,16 @@
 #include <memory>
 #include <mutex>
 #include <string>
+#include <thread>
+#include <vector>
```

Comment:
Per comments below, if you allow setting parameters in the writer class, then we only need a couple of helper functions.